### ISSUE TITLE
Use database replicas for read requests

### ISSUE THREAD

**Issue by gravitystorm** [openstreetmap/openstreetmap-website#2634]
We realised in #2563 that we needed to update our controller tests, so that they use the DatabaseSelector middleware in the test suite and pick up any cases of trying to write to the database during read requests. Now that the refactoring is done I thought I would revisit the multiple databases setup.

However, last week I found a new problem when running the test suite, and it's taken many hours to narrow down the cause (since it lies somewhere deep deep within ActiveRecord, database connections and rails middleware) into a simplified test case, and even then I'm still not sure what's causing it our how to fix/workaround it. I think it's just in the test suite since the rails app appears to work fine for me, and even throws the right error in the right places. But until the problem with the test suite is sorted, this is on hold.

[I've reported the problem I found to the rails team](https://github.com/rails/rails/issues/39449), maybe someone there can point us in the right direction! This issue here is just to share progress with everyone and avoid anyone else having to duplicate the work I've done. If you want to get involved, you can see it (not) working on my [`multiple_databases_2`](https://github.com/gravitystorm/openstreetmap-website/tree/multiple_databases_2) branch, or have a look at the testcase app linked from the rails issue.

**Comment by gravitystorm** [comment]
Great, the bug has already been fixed (I didn't find an existing ticket on the matter) so it'll be in the 6.0.4 release. Just need to wait for that.

**Comment by tomhughes** [comment]
Having just reviewed the old ticket I'm a bit confused...

If it automatically switches to using the primary database reads if you have done a recent write what is all the stuff about the HTTP verb deciding which database to use? Surely attempts to read go to the read database (unless you have written recently) and attempts to write go to the write database?

Is the story that each action has to use a single database and the timeout is actually about sending read actions to a write database if there has been a recent write action?

**Comment by gravitystorm** [comment]
> Is the story that each action has to use a single database and the timeout is actually about sending read actions to a write database if there has been a recent write action?

If there's been a recent write action in the same session, yes. So if you e.g. save your profile then for the next two seconds, all your read requests are routed to the write database. After that, your a session goes back to the normal get/head => read database, everything else => write database.

We can override the default connection within a controller action (e.g. updating the status of a message when it's viewed) but by default the connections are automatically chosen based on the http verb.

See https://github.com/rails/rails/blob/7ddf2b7c0c8ff97b0a50e963b24bf2b82950988d/activerecord/lib/active_record/middleware/database_selector.rb#L58-L71 and https://github.com/rails/rails/blob/7ddf2b7c0c8ff97b0a50e963b24bf2b82950988d/activerecord/lib/active_record/middleware/database_selector/resolver.rb#L34-L40

**Comment by tomhughes** [comment]
Ah right, so the key thing I was missing was that it stored the last write time in the user's session.

I was assuming it was just a global variable that would work across sessions, but not across passenger instances let alone multiple servers.

**Comment by gravitystorm** [comment]
> Just need to wait for that.

In the meantime I've been testing against the 6-0-stable branch of rails, and it turns out that we make a surprising number of writes during GET requests! Way more than I expected. Some of these are reasonable (like updating the "message_read" when viewing messages, or the "needs_view" flag for user_blocks). Some are more dubious but not super important, like resending the registration confirmation email being via GET. Some are tricky, like how our signin system uses a GET request on a redirect to complete the signin process, or the magical mystical I-certainly-don't-understand-this rigmarole of creating tokens during OAuth GET requests. And then some I think can be definitely refactored, like the admin links to suspend or delete users all working via GET requests.

**Comment by mmd-osm** [comment]
Pardon my ignorance. Now that we’re deep down the rabbit hole I somehow lost track what was the original motivation for using a read replica. Is this a „because we can“ type of thing or was it meant to address some actual issue, like too much load on karm (which we don’t seem to have)?

Is our replica prepared for this, including network bandwidth, latency and system availability?

**Comment by tomhughes** [comment]
Well we wouldn't use this at AMS where there is a high performance local database.

We might use it in York so that reads can go to the local slave while write still go to Amsterdam.

**Comment by gravitystorm** [comment]
> Is this a „because we can“ type of thing

For me, mostly yes. It's really easy to switch in rails, and we already have cgimap reading from replica databases so it seems to make sense for other read-only things to come from the replicas too. And even if karm isn't loaded today, our master databases are still really stonkingly expensive servers, so scaling reads without having to scale the primary database will come in useful at some point (e.g. maybe save a few euro on the cost of the next generation master db).

The rabbit hole with the code refactoring is one of our own making, really, rather than being directly related to the database switching. But over the years I've found that fixing some aspects (e.g. bringing our controller tests up to date) often comes in useful in other situations further down the road. So I'm not concerned if the amount of work here seems disproportionate to the matter at hand.

**Comment by gravitystorm** [comment]
At this point I'm not even sure if 6.0.4 will ever be released. Depending on how #3021 goes, we might end up upgrading to 6.1.x anyway before 6.0.4 appears.

**Comment by mmd-osm** [comment]
> we already have cgimap reading from replica databases so it seems to make sense for other read-only things to come from the replicas too

While cgimap can have different settings for read only and update operations, it’s currently not configured that way, i.e. cgimap doesn’t use read replicas at this time.

**Comment by tomhughes** [comment]
That's true, but we have used it that way in the past and we might well in the future - we are likely to be moving to having two "snap" class servers in each of Amsterdam and Dublin at which point we might well switch back to that model.
